### PR TITLE
Use memoization to optimize the fibonacci calculation

### PR DESCRIPTION
**Pull request by iron768**
Use dynamic programming to optimize fib calculations.

### CODE REVIEW

File: src/main/java/org/fibsters/FibCalcComputeEngineImpl.java [hunk 1]
```diff
@@ -112,12 +119,28 @@ public void run() {
         }
 
         if (total > 1) {
-            fibonacci[1] = calculateNthFibonacci(startIndex + 1);
+            //1 1 2 3 5 8 13 21 34 55 89 144 233 377 610 987 1597 2584 4181 6765 10946 17711 28657 46368 75025 121393 196418 317811 514229
+            Map<Integer, Integer> memoizationMap = new HashMap<>();
+
+            // Store the first two Fibonacci numbers directly
+            memoizationMap.put(0, calculateNthFibonacci(startIndex));
+            memoizationMap.put(1, calculateNthFibonacci(startIndex + 1));
```
Comment (ndg8743) [inline src/main/java/org/fibsters/FibCalcComputeEngineImpl.java:127]: Hashmap go brrrr

File: src/main/java/org/fibsters/ComputeJobPoolImpl.java [hunk 2]
```diff
@@ -96,6 +96,8 @@ private void processJob(ComputeJob job) {
                     multiJob.getFibSpiralCE().saveBuffer();
 
                     multiJob.setStatus(ComputeJobStatus.COMPLETED);
+
+                    this.currentJob = null;
```
Comment (lettucegoblin) [inline src/main/java/org/fibsters/ComputeJobPoolImpl.java:100]: Good now it won't get stuck on subsequent multipart jobs

File: src/main/java/org/fibsters/FibCalcComputeEngineImpl.java [hunk 3]
```diff
@@ -12,12 +16,15 @@ public class FibCalcComputeEngineImpl implements FibCalcComputeEngine {
     private InputPayload inputPayload; // used for reference potentially(has uuid)
     private int chunk;
     private OutputPayloadImpl outputPayload;
+    private static Map<Integer, Integer> memoizationMap; // cache
```
Comment (lettucegoblin) [inline src/main/java/org/fibsters/FibCalcComputeEngineImpl.java:19]: Excellent use of caching

**Review by lettucegoblin — APPROVED** [review]
We'll need to automate the test but it's clearly many orders of magnitude faster

File: src/main/java/org/fibsters/FibCalcComputeEngineImpl.java (hunk 3 above)
Comment (ndg8743) [reply src/main/java/org/fibsters/FibCalcComputeEngineImpl.java:19]: image

Comment (ndg8743) [reply src/main/java/org/fibsters/FibCalcComputeEngineImpl.java:19]: image

**Pull request merged**